### PR TITLE
Add permissions, metrics, and metadata attributes to Data Prepper dyn…

### PR DESCRIPTION
**Pull request by graytaylor0**
…amodb source documentation

### Description
This change adds some important information for using the `dynamodb` source in Data Prepper, including metadata attributes that are added to Events, a minimum permissions policy, and metrics that the dynamodb source emits.

### Checklist
- [x] By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and subject to the [Developers Certificate of Origin](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: _data-prepper/pipelines/configuration/sources/dynamo-db.md
Comment (dlvenable) [inline _data-prepper/pipelines/configuration/sources/dynamo-db.md:99]: I think it would be ideal to consistently use `DynamoDB` instead of `DDB`. It makes it easier for the reader.

File: _data-prepper/pipelines/configuration/sources/dynamo-db.md
Comment (dlvenable) [inline _data-prepper/pipelines/configuration/sources/dynamo-db.md:164]: It might be useful to add a paragraph here that tells users they can remove Sid `allowReadFromStream` when doing an export only. And also remove the correct Sid's (`allowRunExportJob`, `allowCheckExportjob`, `allowReadAndWriteToS3ForExport`)  when streaming only.

File: _data-prepper/pipelines/configuration/sources/dynamo-db.md
Comment (dlvenable) [inline _data-prepper/pipelines/configuration/sources/dynamo-db.md:118]: Would the user need `DescribeTable` when using only streams? If so, maybe we could split this out. See my comment below about hinting to the user about reducing permissions when using only export or only stream.

We could also create a GitHub issue and follow-up with it. It needn't hold up the PR if it will take some effort.

**Review by dlvenable — CHANGES_REQUESTED** [review]

File: _data-prepper/pipelines/configuration/sources/dynamo-db.md
Comment (graytaylor0) [reply _data-prepper/pipelines/configuration/sources/dynamo-db.md:118]: I just checked the code and looks like DescribeTable is required for streams as well

**Review by dlvenable — APPROVED** [review]
This looks good. Thank you!

File: _data-prepper/pipelines/configuration/sources/dynamo-db.md
Comment (Naarcha-AWS) [inline _data-prepper/pipelines/configuration/sources/dynamo-db.md:93]:
```suggestion
The following metadata will be added to each rvent that is processed by the `dynamodb` source. These metadata attributes can be accessed with the [expression syntax `getMetadata` function](https://opensearch.org/docs/latest/data-prepper/pipelines/expression-syntax/#getmetadata).
```

File: _data-prepper/pipelines/configuration/sources/dynamo-db.md
Comment (Naarcha-AWS) [inline _data-prepper/pipelines/configuration/sources/dynamo-db.md:94]:
```suggestion

```

File: _data-prepper/pipelines/configuration/sources/dynamo-db.md
Comment (Naarcha-AWS) [inline _data-prepper/pipelines/configuration/sources/dynamo-db.md:96]:
```suggestion
* `primary_key` - The primary key of the DynamoDB item. For tables that only contain a partition key, this value gives the partition key. For tables that contain both a partition and sort key, the` primary_key` attribute will be equal to the partition and sort key, separated by a `|`, for example `partition_key|sort_key`.
```

File: _data-prepper/pipelines/configuration/sources/dynamo-db.md
Comment (Naarcha-AWS) [inline _data-prepper/pipelines/configuration/sources/dynamo-db.md:98]:
```suggestion
* `sort_key` - The sort key of the DynamoDB item. This will be null if the table does not have a sort key.
```

File: _data-prepper/pipelines/configuration/sources/dynamo-db.md
Comment (Naarcha-AWS) [inline _data-prepper/pipelines/configuration/sources/dynamo-db.md:99]:
```suggestion
* `dynamodb_timestamp` - The timestamp of the DynamoDB item. This will be the export time for export items and the DynamoDB stream event time for stream items. This timestamp is used by sinks to emit an `EndtoEndLatency` metric for DynamoDB stream events, which tracks the latency between a change in the DynamoDB table and that change getting applied to the sink.
```

File: _data-prepper/pipelines/configuration/sources/dynamo-db.md
Comment (Naarcha-AWS) [inline _data-prepper/pipelines/configuration/sources/dynamo-db.md:100]:
```suggestion
* `document_version` - Uses the `dynamodb_timestamp` to modify break ties between stream items that are received in the same second. Recommend for use with the `opensearch` sink's `document_version` setting.
```

File: _data-prepper/pipelines/configuration/sources/dynamo-db.md
Comment (Naarcha-AWS) [inline _data-prepper/pipelines/configuration/sources/dynamo-db.md:101]:
```suggestion
* `opensearch_action` - A default value for mapping DynamoDB event actions to OpenSearch actions. This action will be `index` for export items, and `INSERT` or `MODIFY` for stream events, and `REMOVE` stream events when the OpenSearch action is `delete`.
```

File: _data-prepper/pipelines/configuration/sources/dynamo-db.md
Comment (Naarcha-AWS) [inline _data-prepper/pipelines/configuration/sources/dynamo-db.md:108]:
```suggestion
The following shows the minimum required permissions for running DynamoDB as a source:
```

File: _data-prepper/pipelines/configuration/sources/dynamo-db.md
Comment (Naarcha-AWS) [inline _data-prepper/pipelines/configuration/sources/dynamo-db.md:174]:
```suggestion
When performing an export, `"Sid": "allowReadFromStream"` section is not required. If only reading from DynamoDB streams, the 
```

File: _data-prepper/pipelines/configuration/sources/dynamo-db.md
Comment (Naarcha-AWS) [inline _data-prepper/pipelines/configuration/sources/dynamo-db.md:179]:
```suggestion
The `dynamodb` source includes the following metrics:
```

File: _data-prepper/pipelines/configuration/sources/dynamo-db.md
Comment (Naarcha-AWS) [inline _data-prepper/pipelines/configuration/sources/dynamo-db.md:183]:
```suggestion
* `exportJobSuccess` - The number of export jobs that have been submitted successfully.
```

File: _data-prepper/pipelines/configuration/sources/dynamo-db.md
Comment (Naarcha-AWS) [inline _data-prepper/pipelines/configuration/sources/dynamo-db.md:184]:
```suggestion
* `exportJobFailure` - The number of export job submission attempts that have failed.
```

File: _data-prepper/pipelines/configuration/sources/dynamo-db.md
Comment (Naarcha-AWS) [inline _data-prepper/pipelines/configuration/sources/dynamo-db.md:185]:
```suggestion
* `exportS3ObjectsTotal` - The total number of export data files found in S3 for an export.
```

File: _data-prepper/pipelines/configuration/sources/dynamo-db.md
Comment (Naarcha-AWS) [inline _data-prepper/pipelines/configuration/sources/dynamo-db.md:186]:
```suggestion
* `exportS3ObjectsProcessed` - The total number of export data files that have been processed successfully from S3.
```

File: _data-prepper/pipelines/configuration/sources/dynamo-db.md
Comment (Naarcha-AWS) [inline _data-prepper/pipelines/configuration/sources/dynamo-db.md:187]:
```suggestion
* `exportRecordsTotal` - The total number of records found in the entire export.
```

File: _data-prepper/pipelines/configuration/sources/dynamo-db.md
Comment (Naarcha-AWS) [inline _data-prepper/pipelines/configuration/sources/dynamo-db.md:188]:
```suggestion
* `exportRecordsProcessed` - The total number of export records that have been processed successfully.
```

File: _data-prepper/pipelines/configuration/sources/dynamo-db.md
Comment (Naarcha-AWS) [inline _data-prepper/pipelines/configuration/sources/dynamo-db.md:189]:
```suggestion
* `exportRecordsProcessingErrors` - The number of export record processing errors.
```

File: _data-prepper/pipelines/configuration/sources/dynamo-db.md
Comment (Naarcha-AWS) [inline _data-prepper/pipelines/configuration/sources/dynamo-db.md:190]:
```suggestion
* `changeEventsProcessed` - The number of change events processed from DynamoDB streams.
```

File: _data-prepper/pipelines/configuration/sources/dynamo-db.md
Comment (Naarcha-AWS) [inline _data-prepper/pipelines/configuration/sources/dynamo-db.md:191]:
```suggestion
* `changeEventsProcessingErrors` - The number of processing errors for change events from DynamoDB streams.
```

File: _data-prepper/pipelines/configuration/sources/dynamo-db.md
Comment (Naarcha-AWS) [inline _data-prepper/pipelines/configuration/sources/dynamo-db.md:192]:
```suggestion
* `shardProgress` - The incremented shard progress when DynamoDB streams are being read correctly. If this is `0` for a certain period of time, then there is a problem with the pipeline that has streams enabled.
```

**Review by Naarcha-AWS — APPROVED** [review]

File: _data-prepper/pipelines/configuration/sources/dynamo-db.md
Comment (natebower) [inline _data-prepper/pipelines/configuration/sources/dynamo-db.md:100]: "This action will be `index` for export items, `INSERT` or `MODIFY` for stream events, and `REMOVE` when the OpenSearch action is `delete`"? (The second sentence needs a little revision for consistency.)

File: _data-prepper/pipelines/configuration/sources/dynamo-db.md
Comment (natebower) [inline _data-prepper/pipelines/configuration/sources/dynamo-db.md:101]:
```suggestion
* `dynamodb_event_name` - The exact event type for the item. Will be `null` for export items and either `INSERT`, `MODIFY`, or `REMOVE` for stream events.
```

File: _data-prepper/pipelines/configuration/sources/dynamo-db.md
Comment (natebower) [inline _data-prepper/pipelines/configuration/sources/dynamo-db.md:102]:
```suggestion
* `table_name` - The name of the DynamoDB table that an event came from.
```

File: _data-prepper/pipelines/configuration/sources/dynamo-db.md
Comment (natebower) [inline _data-prepper/pipelines/configuration/sources/dynamo-db.md:107]: Can "shows" just be "are"?

File: _data-prepper/pipelines/configuration/sources/dynamo-db.md
Comment (natebower) [inline _data-prepper/pipelines/configuration/sources/dynamo-db.md:173]:
```suggestion
When performing an export, the `"Sid": "allowReadFromStream"` section is not required. If only reading from DynamoDB streams, the 
```

File: _data-prepper/pipelines/configuration/sources/dynamo-db.md
Comment (natebower) [inline _data-prepper/pipelines/configuration/sources/dynamo-db.md:174]:
```suggestion
`"Sid": "allowReadAndWriteToS3ForExport"`, `"Sid": "allowCheckExportjob"`, and ` "Sid": "allowRunExportJob"` sections are not required.
```

File: _data-prepper/pipelines/configuration/sources/dynamo-db.md
Comment (natebower) [inline _data-prepper/pipelines/configuration/sources/dynamo-db.md:178]:
```suggestion
The `dynamodb` source includes the following metrics.
```

File: _data-prepper/pipelines/configuration/sources/dynamo-db.md
Comment (natebower) [inline _data-prepper/pipelines/configuration/sources/dynamo-db.md:182]: The list items should all be separated from their descriptions by either a colon (`partition_key`: The) or an en dash (`partition_key` -- The).

File: _data-prepper/pipelines/configuration/sources/dynamo-db.md
Comment (natebower) [inline _data-prepper/pipelines/configuration/sources/dynamo-db.md:184]: Can "for an export" be removed?

File: _data-prepper/pipelines/configuration/sources/dynamo-db.md
Comment (natebower) [inline _data-prepper/pipelines/configuration/sources/dynamo-db.md:186]:
```suggestion
* `exportRecordsTotal` - The total number of records found in the export.
```

File: _data-prepper/pipelines/configuration/sources/dynamo-db.md
Comment (natebower) [inline _data-prepper/pipelines/configuration/sources/dynamo-db.md:191]:
```suggestion
* `shardProgress` - The incremented shard progress when DynamoDB streams are being read correctly. This being `0` for a certain period of time indicates a problem with the pipeline that has streams enabled.
```

Comment (natebower) [inline _data-prepper/pipelines/configuration/sources/dynamo-db.md:191]: Should we specify what we mean by "a certain period of time", or do we mean something like "for any length of time"?

**Review by natebower — COMMENTED** [review]
@graytaylor0 @Naarcha-AWS Please see my comments and changes and let me know if you have any questions. Thanks!

File: _data-prepper/pipelines/configuration/sources/dynamo-db.md
Comment (Naarcha-AWS) [reply _data-prepper/pipelines/configuration/sources/dynamo-db.md:99]: Looked into this a bit more. At first, I thought that there's a chance that you could receive a request within milliseconds of each other, which wouldn't occur simultaneously. However, It looks like the DynamoDB timestamp format ends at the second counter (see https://aws.amazon.com/blogs/database/working-with-date-and-timestamp-data-types-in-amazon-dynamodb/). With that in mind, I think simultaneously is easier to understand.

Comment (Naarcha-AWS) [reply _data-prepper/pipelines/configuration/sources/dynamo-db.md:99]: Let me know if modified suggestions works for you @natebower

File: _data-prepper/pipelines/configuration/sources/dynamo-db.md
Comment (Naarcha-AWS) [inline _data-prepper/pipelines/configuration/sources/dynamo-db.md:100]:
```suggestion
* `opensearch_action` - A default value for mapping DynamoDB event actions to OpenSearch actions. This action will be `index` for export items,`INSERT` or `MODIFY` for stream events, and `REMOVE` for deleted stream events. 
```

File: _data-prepper/pipelines/configuration/sources/dynamo-db.md
Comment (natebower) [reply _data-prepper/pipelines/configuration/sources/dynamo-db.md:99]: @Naarcha-AWS LGTM once the hyphen is changed to either an en dash or colon.

File: _data-prepper/pipelines/configuration/sources/dynamo-db.md
Comment (Naarcha-AWS) [reply _data-prepper/pipelines/configuration/sources/dynamo-db.md:100]: @natebower: ?

File: _data-prepper/pipelines/configuration/sources/dynamo-db.md
Comment (Naarcha-AWS) [inline _data-prepper/pipelines/configuration/sources/dynamo-db.md:107]:
```suggestion
The following are the minimum required permissions for running DynamoDB as a source:
```

File: _data-prepper/pipelines/configuration/sources/dynamo-db.md
Comment (Naarcha-AWS) [reply _data-prepper/pipelines/configuration/sources/dynamo-db.md:191]: "For any significant length of time?"

File: _data-prepper/pipelines/configuration/sources/dynamo-db.md
Comment (natebower) [reply _data-prepper/pipelines/configuration/sources/dynamo-db.md:100]: @Naarcha-AWS LGTM

File: _data-prepper/pipelines/configuration/sources/dynamo-db.md
Comment (Naarcha-AWS) [inline _data-prepper/pipelines/configuration/sources/dynamo-db.md:191]:
```suggestion
* `shardProgress`: The incremented shard progress when DynamoDB streams are being read correctly. This being`0` for any significant amount of time means there is a problem with the pipeline that has streams enabled.
```

File: _data-prepper/pipelines/configuration/sources/dynamo-db.md
Comment (Naarcha-AWS) [reply _data-prepper/pipelines/configuration/sources/dynamo-db.md]: I think its really dependent on the user's configuration and what their own expected shard progress time is.

Comment (natebower) [reply _data-prepper/pipelines/configuration/sources/dynamo-db.md]: Ok, then "For any significant length of time" works.

**Pull request merged**
